From f3ae441a5ff9d27a3dd8223d55a7367ae519a0c8 Mon Sep 17 00:00:00 2001
From: Matthew Yeazel <yeazelm@amazon.com>
Date: Wed, 7 Jun 2023 13:34:49 -0700
Subject: [PATCH] Update shutdown code to work for Bottlerocket

Bottlerocket doesn't have /bin/sh which makes the call system() fail
with an unexpected error code. This results in the guest shutdown and
reboot calls via vmtoolsd returning as if the request was successful but
the call actually failed completely. This commit replaces the current
call to system() with g_spawn_command_line_sync() which should be a drop
in replacement for system().

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>
[jpculp: port to open-vm-tools 13.1.0]
Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>
---
 lib/system/Makefile.am   |  3 +++
 lib/system/systemLinux.c | 26 +++++++++++++++++++-------
 2 files changed, 22 insertions(+), 7 deletions(-)

diff --git a/lib/system/Makefile.am b/lib/system/Makefile.am
index 52b1a1dd..d2613c43 100644
--- a/lib/system/Makefile.am
+++ b/lib/system/Makefile.am
@@ -17,5 +17,8 @@
 
 noinst_LTLIBRARIES = libSystem.la
 
+libSystem_la_CPPFLAGS =
+libSystem_la_CPPFLAGS += @GLIB2_CPPFLAGS@
+
 libSystem_la_SOURCES =
 libSystem_la_SOURCES += systemLinux.c
diff --git a/lib/system/systemLinux.c b/lib/system/systemLinux.c
index c7fc6146..c4107a1b 100644
--- a/lib/system/systemLinux.c
+++ b/lib/system/systemLinux.c
@@ -55,6 +55,7 @@
 #include <sys/ioctl.h>
 #include <net/if.h>
 #include <sys/ioctl.h>
+#include <glib.h>
 
 #if defined sun || defined __APPLE__
 #   include <utmpx.h>
@@ -780,10 +781,10 @@ SNEForEachCallback(const char *key,     // IN: environment variable
  *
  * SystemRunShutdownCommand --
  *
- *   Run a reboot or shutdown command using system() and log a warning if it fails.
+ *   Run a reboot or shutdown command and log a warning if it fails.
  *
  * Return value:
- *    Returns the return value of the system() call.
+ *    Returns the exit status of the command, or -1 on exec failure.
  *
  * Side effects:
  *    None.
@@ -795,17 +796,28 @@ static int
 SystemRunShutdownCommand(Bool reboot,     // IN: "reboot or shutdown" flag
                          const char *cmd) // IN: "command to run"
 {
-   int status;
+   int status = 0;
+   Bool success;
+
+   success = g_spawn_command_line_sync(cmd,
+                           NULL,
+                           NULL,
+                           &status,
+                           NULL);
+
+   if (!success) {
+      Warning("%s: %s command \"%s\": failed to execute\n",
+              __func__, reboot ? "reboot" : "shutdown", cmd);
+      return -1;
+   }
 
-   if ((status = system(cmd)) != 0) {
+   if (status != 0) {
       char cmsg[128];
 
       Str_Snprintf(cmsg, sizeof(cmsg), "%s: %s command \"%s\"",
                    __func__, reboot ? "reboot" : "shutdown", cmd);
 
-      if (status == -1) {
-         Warning("%s: failed. Error: %s\n", cmsg, Err_Errno2String(errno));
-      } else if (WIFEXITED(status)) {
+      if (WIFEXITED(status)) {
          Warning("%s exited. status: %d\n", cmsg, WEXITSTATUS(status));
       } else if (WIFSIGNALED(status)) {
          Warning("%s terminated by signal: %d\n", cmsg, WTERMSIG(status));
-- 
2.37.1 (Apple Git-137.1)

